src/tests/mount_injection.c

```diff
@@ -0,0 +1,279 @@
+/* mount_injection.c
+ *
+ * Copyright © 2018
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2, as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#define _GNU_SOURCE
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <lxc/lxccontainer.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "lxctest.h"
+#include "utils.h"
+
+#define NAME "mount_injection_test"
+#define TEMPLATE P_tmpdir"/mount_injection_XXXXXX"
+
+struct mountinfo_strings {
+	const char *mount_root;
+	const char *mount_point;
+	const char *fstype;
+	const char *mount_source;
+};
+
+static int comp_field(char *line, const char *str, int nfields)
+{
+	char *p, *p2;
+	int i, ret;
+
+	if(!line)
+		return -1;
+
+	if (!str)
+		return 0;
+
+	for (p = line, i = 0; p && i < nfields; i++)
+		p = strchr(p + 1, ' ');
+	if (!p)
+		return -1;
+	p2 = strchr(p + 1, ' ');
+	if (p2)
+		*p2 = '\0';
+	ret = strcmp(p + 1, str);
+	if (p2)
+		*p2 = ' ';
+	return ret;
+}
+
+static int find_in_proc_mounts(void *data)
+{
+	char buf[LXC_LINELEN];
+	FILE *f;
+	struct mountinfo_strings *strs = (struct mountinfo_strings *) data;
+
+	f = fopen("/proc/self/mountinfo", "r");
+	if (!f)
+		return 0;
+	while (fgets(buf, LXC_LINELEN, f)) {
+		fprintf(stderr, "next loop\n");
+		if (comp_field(buf, strs->mount_root, 3) == 0 && comp_field(buf, strs->mount_point, 4) == 0) {
+			char *buf2 = strchr(buf, '-');
+			if (comp_field(buf2, strs->fstype, 1) == 0 && comp_field(buf2, strs->mount_source, 2) == 0) {
+				fclose(f);
+				fprintf(stderr, "OK for %s\n", strs->mount_point);
+				fprintf(stdout, "OK");
+				_exit(EXIT_SUCCESS);
+			}
+		}
+	}
+	fclose(f);
+	fprintf(stderr, "ER for %s\n", strs->mount_point);
+	fprintf(stdout, "ER");
+	_exit(EXIT_FAILURE);
+}
+
+static int check_containers_mountinfo(struct lxc_container *c, struct mountinfo_strings *d)
+{
+	int pipefd[2];
+	pid_t pid;
+	char resp[3];
+	int ret = -1;
+	lxc_attach_options_t attach_options = LXC_ATTACH_OPTIONS_DEFAULT;
+
+	ret = pipe(pipefd);
+	if (ret < 0) {
+		fprintf(stderr, "Pipe failed\n");
+		goto out_pipe;
+	}
+	attach_options.stdout_fd = pipefd[1];
+	ret = c->attach(c, find_in_proc_mounts, d, &attach_options, &pid);
```

Comment:
Oh, I think this can be done way simpler. :) You need to wait for the process to exit and retrieve it's exit status at which point you'll know whether `find_in_proc_mount()` succeeded or not. So all the `pipe()` based logic can go. What you need is:
```
ret = c->attach(c, find_in_proc_mounts, d, &attach_options, &pid);
if (ret < 0)
        /* handle error */
ret = wait_for_pid(pid);
if (ret < 0)
        /* your function failed */

/* your function succeeded */
```